### ISSUE TITLE
[AMD][kernel][Tier 5] FlyDSL fused linear-head + cross-entropy bwd for training memory win (split from #33)

### ISSUE THREAD

**Issue by sunxxuns** [amdpilot-org/FlyDSL#37]
## Goal

Use FlyDSL to generate the **fused linear-head + cross-entropy bwd
kernel** for training: backward — fused dh + dW_lm_head directly from saved (targets, lse, h, W_lm_head); again no logits materialisation. One of 2 sibling issues splitting the
former FlyDSL #33 — the joint fwd+bwd umbrella was over-scoped for
Stage 0.

Backward path — needs fwd's saved (lse, targets) tuple as input.

Reuses cached **`amd/Llama-3.3-70B-Instruct-MXFP4-Preview`** (vocab
128k) and **`Qwen/Qwen3-32B`** (vocab 152k) for cross-vocab-size
coverage.

## Hardware / workload

```
Hardware:    1 × MI355X (kernel-level)
Shape:       hidden=8192, vocab=128k, B*S=4096 tokens
             (Llama-3.3-70B training shape at micro_batch=2 × S=2048)
Baseline:    F.linear's autograd backward (which has the same logits-materialisation cost on the reverse pass)
Target:      fused bwd ms <= 1.2 × baseline_linear_only_ms (the
             reduction/CE add itself is negligible vs the GEMM);
             peak HBM working set < 0.2 × baseline (no logits
             materialisation).
```

## Scope (single kernel)

This issue ships exactly one kernel.

1. Add `kernels/fused_linear_cross_entropy_bwd.py` — the
   GEMM + streaming epilogue (forward) OR symmetric reverse (backward,
   needs the saved `(lse, targets, h, W_lm_head)` tuple).
2. Wire into a small autograd helper in `python/flydsl/ops/loss.py` —
   forward variant fills the `forward` slot; backward fills
   `backward`. (Both halves together produce the complete
   `torch.autograd.Function`; ship-order doesn't matter — the
   half that lands second completes the wiring.)
3. Parity test in `tests/kernels/test_fused_linear_ce_bwd.py`
   vs the PyTorch eager reference (forward: assert loss matches +
   peak HBM working set < 0.2× baseline; backward: assert dh + dW
   gradients match the eager autograd reference).
4. Memory benchmark in
   `tests/kernels/bench_fused_linear_ce_bwd.py` reporting
   peak HBM working set + `kernel_ms` for the Llama-3.3-70B shape.

## Acceptance

- Parity test passes (max_abs_diff < 1e-3 on loss / < 5e-3 on dh/dW
  for backward).
- Peak HBM working set < 0.2 × baseline (the logits tensor must not
  materialise on this half).
- `kernel_ms <= 1.1 × baseline_linear_only_ms` (no regression vs
  the linear-only step).

## Why this is a single-kernel issue

The joint fwd+bwd umbrella was in amdpilot-org/FlyDSL#33. Splitting
gives Stage 0 a single clean benchmark metric per issue (forward
returns a loss; backward returns dh/dW grads — two different test
harnesses, one kernel each).

Sibling: the other of fwd / bwd.

**Comment by jhinpan** [comment]
**AMDPilot — accumulated progress on this issue**

_Auto-generated from successive trials. Latest update: 2026-05-23T14:35:49Z by `issue-FlyDSL-37-r13`._

### Verified infra fixes (apply during Stage 0)

- Patch test_harness.py to fix NameError ('k' vs 'key') in rocm_mem_mb() preventing STAGE0_RESULT_JSON generation
- Fix malformed f-strings with unmatched brackets in shell commands causing GitHub API SyntaxErrors


### Working approaches and their best metric

_(none yet — when a trial measurably improves the metric, the change lands here.)_


### Approaches that did NOT help — don't retry these

- Running raw PyTorch CUDA sanity checks instead of executing the baseline benchmark
- Redundant grep/cat scans of /workspace/FlyDSL/tests and /workspace/FlyDSL/python for kernel patterns instead of running the profiler
- Re-reading test_vec_add.py after repository structure was already confirmed in step 13
- Empty shell command execution causing JSON validation errors in escalation stage
- Chunked softmax backward pass in bfloat16 is numerically unstable (error ~1.0)
- Single-chunk fused kernel implementation still fails (error ~6.35), ruling out chunk-boundary logic as sole cause
- Manually constructing alias shims (test_meta.py, _mlir/__init__.py) to resolve imports instead of running benchmark
- Rebuilding MLIR components and re-reading CLAUDE.md despite existing context
- Attempting to symlink _mlir directory which triggers PyGlobals assertion failure on import


### Open ideas to try next

- Enforce a hard stop on file-system exploration after a baseline-run nudge is issued
- Explicitly install missing flydsl.expr and _mlir bindings into /opt/venv
- Enforce a hard gate in the Orchestrator to block grep/cat tools until baseline_profile.py is created or executed
- Update Orchestrator parser to extract STAGE0_RESULT_JSON from stdout before checking artifact files
- Optimize fused kernel implementation to exceed eager execution speed (currently 0.43x)
- Implement 'script-dedup' guard in Shell tool to force iterative debugging over redundant file creation
- Enforce mandatory profiling nudge to save traces to /workspace/traces/ before allowing new script generation
- Investigate numerical stability fixes for bfloat16 softmax backward or switch to fp32 accumulation
- Enforce hard gate in Orchestrator to block grep/cat tools until baseline_profile.py is executed
- Implement script-dedup guard in Shell tool to force iterative debugging over redundant file creation
- Add mandatory profiling nudge to save traces to /workspace/traces/ before allowing new script generation


_amdpilot bot. Edit between the markers to add/correct fixes; the next update will preserve your changes. To disable updates on this issue, replace the body with the marker line only._

**Comment by sunxxuns** [comment]
## Closing as not tractable by the current autonomous agent (round 2, 2026-05-23)

Chronic-zero-win lineage: 4+ retries in 12 h with 0 completions. Triage: fused linear-head + CE backward — Liger-style memory fusion needs careful schedule.

The agent (Kimi-K2.6) cannot make iterative progress on this scope alone; every
retry lands on the same blocker. Closing so the fleet stops burning GPU cycles.

**To re-open**: decompose into a single-shape sweep or single-kernel-with-measurable-metric task, or assign to a human kernel engineer.
